pkg/controllers/corecluster/corecluster_controller.go

```diff
@@ -81,38 +87,89 @@ func (r *CoreClusterController) SetupWithManager(mgr ctrl.Manager) error {
 	logger := log.FromContext(ctx).WithName(controllerName)
 	logger.Info("Reconciling core cluster")
 	defer logger.Info("Finished reconciling core cluster")
 
+	var err error
+
+	co, err := r.GetOrCreateClusterOperator(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get or create cluster operator: %w", err)
+	}
+
+	// Populate conditions from the existing cluster operator status.
+	conditions := &co.Status.Conditions
+
+	defer func() {
+		// This function runs before every return, to ensure the controller conditions
+		// for that codepath are correctly synced to the cluster operator status.
+		if syncErr := r.SyncControllerConditions(ctx, co, controllerName, conditions,
+			[]string{CoreClusterControllerAvailableCondition, CoreClusterControllerDegradedCondition}); syncErr != nil {
+			if err != nil {
+				// The sync of the controller conditions failed but there is also a non nil reconcile error.
+				// Log the sync error and pass down the reconciler one.
+				logger.Error(syncErr, "failed to sync cluster operator status")
+			} else {
+				// If there is no reconcile error, but there is a sync error, then we want to propagate the
+				// sync error to be returned by the Reconcile function, so that it will cause an immediate requeue/retry.
+				err = syncErr
+			}
+		}
+	}()
+
 	ocpInfrastructureName, err := getOCPInfrastructureName(r.Infra)
 	if err != nil {
+		operatorstatus.SetCondition(conditions, CoreClusterControllerDegradedCondition, configv1.ConditionTrue,
+			operatorstatus.ReasonSyncFailed, fmt.Sprintf("controller is degraded: %s", err.Error()))
+
+		operatorstatus.SetCondition(conditions, CoreClusterControllerAvailableCondition, configv1.ConditionTrue,
+			operatorstatus.ReasonAsExpected, "controller is available")
+
 		return ctrl.Result{}, fmt.Errorf("failed to obtain infrastructure name: %w", err)
 	}
 
 	cluster, err := r.ensureCoreCluster(ctx, client.ObjectKey{Namespace: r.ManagedNamespace, Name: ocpInfrastructureName}, logger)
 	if err != nil {
+		operatorstatus.SetCondition(conditions, CoreClusterControllerDegradedCondition, configv1.ConditionTrue,
+			operatorstatus.ReasonSyncFailed, fmt.Sprintf("controller is degraded: %s", err.Error()))
+
+		operatorstatus.SetCondition(conditions, CoreClusterControllerAvailableCondition, configv1.ConditionTrue,
+			operatorstatus.ReasonAsExpected, "controller is available")
+
 		return ctrl.Result{}, fmt.Errorf("failed to ensure core cluster: %w", err)
 	}
 
 	if !cluster.DeletionTimestamp.IsZero() {
-		if err := r.SetStatusAvailable(ctx, ""); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
-		}
+		operatorstatus.SetCondition(conditions, CoreClusterControllerDegradedCondition, configv1.ConditionFalse,
+			operatorstatus.ReasonAsExpected, "controller works as expected")
+
+		operatorstatus.SetCondition(conditions, CoreClusterControllerAvailableCondition, configv1.ConditionTrue,
```

Comment:
If the cluster is being deleted, are we really available?

Reply:
That's trickier. I think we will end up disallowing core cluster deletion via VAP/WebHook, but the operator will still keep reconciling without error and will recreate this object.

Either could be ok depending on how we look at things.
If we don't explicitly set it here, we are probably inheriting Available from the previous state.

Ok, now I dropped and updated the test, let's see how we roll with this.

Reply:
It seemed to be ending the reconcile once it detected the deletion timestamp? Is that expected then? That seems unusual given your description that it should continue to reconcile

Reply:
Keep reconciling I mean coming back to a new reconciliation cycle once the object as been fully deleted an perform a recreation.

Bear in mind I haven't touched this logic at all with the PR, so the behaviour is as it was prior to my changes.

One thing we can do here to improve speed of recovery from object deletion is probably to add a requeueAfter to detect the object is being deleted. So we can catch it quickly and recreate.

Reply:
Talked about this out of band. It's cool.